### PR TITLE
Add help message,default value and validation in function:add

### PR DESCRIPTION
**Pull request by Flanker32**
Add explanation of attributes when function:add
Add default value and validation in function:add for string value in interactive mode

### CODE REVIEW

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/template/FunctionSettingTemplate.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+package com.microsoft.azure.maven.function.template;
+
+import com.fasterxml.jackson.annotation.JsonGetter;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonSetter;
+
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class FunctionSettingTemplate {
```
Comment (jdneo) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/template/FunctionSettingTemplate.java]: Add comments to point out which data structure does this class refer to.

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/template/TemplateTriggerTypeDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+package com.microsoft.azure.maven.function.template;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
+
+import java.io.IOException;
+
+public class TemplateTriggerTypeDeserializer extends StdDeserializer<String> {
```
Comment (jdneo) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/template/TemplateTriggerTypeDeserializer.java]: Add comments to explain it's input and output.

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java [hunk 3]
```diff
@@ -319,6 +341,40 @@ protected void preparePackageName() throws MojoFailureException {
...
 
+    protected String getStringInputFromUser(String attributeName, String initValue, FunctionSettingTemplate template) {
+        final String regex = template == null ? null : template.getSettingRegex();
+        final String defaultValue = template == null ? null : template.getDefaultValue();
+        final String prompt = defaultValue == null ?
+            String.format("Enter value for %s: ", attributeName) :
+            String.format("Enter value for %s(Default: %s): ", attributeName, defaultValue);
+        final String errorMessage = (template != null && template.getErrorText() != null) ?
+            TemplateResources.getResource(template.getErrorText()) : DEFAULT_INPUT_ERROR_MESSAGE;
+        final Function<String, Boolean> validator = regex == null ?
```
Comment (jdneo) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java]: Too much `(?:)` in this method, which makes the logic here a little bit confusing.

You can write it like this:

```java 
Function<String, Boolean> validator = StringUtils::isNotEmpty;
if (template != null && StringUtils.isNotEmpty(template.getSettingRegex())) {
    validator = (attribute) -> StringUtils.isNotEmpty(attribute) && attribute.matches(template.getSettingRegex());
}
```

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java [hunk 4]
```diff
@@ -319,6 +341,40 @@ protected void preparePackageName() throws MojoFailureException {
...
+            StringUtils::isNotEmpty :
+            (attribute) -> StringUtils.isNotEmpty(attribute) && attribute.matches(regex);
+
+        if (validator.apply(initValue)) {
+            info(FOUND_VALID_VALUE);
+            return initValue;
+        }
+
+        final Scanner scanner = getScanner();
```
Comment (jdneo) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java]: Move `prompt` and `errorMessage` closer to where they are used

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java [hunk 5]
```diff
@@ -319,6 +341,40 @@ protected void preparePackageName() throws MojoFailureException {
...
+        while (true) {
+            out.printf(prompt);
+            out.flush();
```
Comment (jdneo) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java]: Why do we need to flush manually here?

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java [hunk 6]
```diff
@@ -319,6 +341,40 @@ protected void preparePackageName() throws MojoFailureException {
...
+            try {
+                final String input = scanner.nextLine();
+                if (validator.apply(input)) {
+                    return input;
+                } else if (defaultValue != null && !StringUtils.isNotEmpty(input)) {
```
Comment (jdneo) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java]: `!StringUtils.isNotEmpty(input)` -> StringUtils.isEmpty(input)

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java (hunk 3 above)
Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java]: Done, and move most of them to simple function

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java [hunk 7]
```diff
@@ -158,18 +161,14 @@ protected void doExecute() throws Exception {
...
             final String bindingsJsonStr = IOUtil.toString(is);
             final BindingsTemplate bindingsTemplate = new ObjectMapper()
                 .readValue(bindingsJsonStr, BindingsTemplate.class);
-            info(LOAD_TEMPLATES_DONE);
             return bindingsTemplate.getBindingTemplateByName(type);
         } catch (Exception e) {
-            error(LOAD_TEMPLATES_FAIL);
-            throw e;
+            warning(LOAD_BINDING_TEMPLATES_FAIL);
+            return null;
```
Comment (jdneo) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java]: Why we swallow the error?

Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/AddMojo.java]: BindingTemplate only contains default value and prompt message, addMojo could also work without it, so just swallow this error and show warning here.

**Pull request merged**